soc: silabs: siwx91x: Add siwx91x Power Manager driver

fimohame opened the pull request:
This commit introduces Power Manager driver support for the siwx917 device, enabling both PS4 Standby and PS4 Sleep modes. Additionally, it enables runtime GPIO driver support for the siwx917 device.

fimohame committed:
soc: silabs: siwx91x: Add siwx91x Power Manager driver

This commit enables the Power Manager driver
support for the siwx91x device.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

fimohame committed:
driver: gpio: siwx91x: Add device runtime support for gpio driver

This commit enables the device runtime driver support
for the siwx91x device.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

fimohame committed:
west.yml: Update hal_silabs revision

This commit updates the hal_silabs revision

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

asmellby commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -202,6 +220,24 @@ static int siwx91x_clock_init(const struct device *dev)
 
+#ifdef CONFIG_PM
+	/* Initialize platform-specific configurations. */
+	sli_si91x_platform_init();
```
Since this seems to have more to do with power manager than clocks, is there a good reason to have it in the clock driver instead of `soc.c` using one of the init hooks?

asmellby commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -295,6 +301,34 @@ static inline int gpio_siwx91x_init_port(const struct device *port)
+	case PM_DEVICE_ACTION_RESUME:
+		for (int pin = 0; pin < MAX_PIN_COUNT; ++pin) {
+			if (config->common.port_pin_mask & BIT(pin)) {
+				sl_si91x_gpio_set_pin_direction(config->hal_port, pin,
```
Is pin direction the only configuration that is unretained in sleep mode?

fimohame replied:
Yes

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -24,6 +24,19 @@
 		cpu0: cpu@0 {
 			compatible = "arm,cortex-m4f";
 			reg = <0>;
+			cpu-power-states = <&pstate_ps4_standby &pstate_ps4_sleep>;
```
Should the sleep mode be skipped until all drivers implement power management callbacks properly? The state could still be defined in DTS, just not registered for use by the CPU.

asmellby commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -123,7 +140,7 @@ endif() # CONFIG_BT_SILABS_SIWX91X
   zephyr_compile_definitions(
     SLI_SI91X_ENABLE_OS
-    SL_SI91X_SI917_RAM_MEM_CONFIG=1
+    SL_SI91X_SI917_RAM_MEM_CONFIG=2
```
This configuration should use DTS values to set the correct value 1, 2 or 3, not hard-code one.

asmellby commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -178,7 +200,44 @@ if(CONFIG_SOC_SILABS_SLEEPTIMER)
+    SL_CODE_COMPONENT_POWER_MANAGER=power_manager
+    SL_SI91X_TICKLESS_MODE
+    SL_SLEEP_TIMER
+    SL_SI91X_SI917_RAM_MEM_CONFIG=2
```
Same as above -- 1, 2 or 3 depending on RAM size config.

asmellby commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+
+	if (sl_si91x_power_manager_is_ok_to_sleep()) {
+		/* Device is ready to sleep; proceed with sleep operations. */
+		if (target_power_state == SL_SI91X_POWER_MANAGER_STANDBY) {
```
This could be simplified to `if (state == PM_STATE_RUNTIME_IDLE)`, and the above switch could be removed.

asmellby commented on soc/silabs/silabs_siwx91x/Kconfig.defconfig:
```diff
@@ -13,6 +13,11 @@ configdefault SYS_CLOCK_TICKS_PER_SEC
 	default 128 if !TICKLESS_KERNEL && SILABS_SLEEPTIMER_TIMER
 	default 1024 if SILABS_SLEEPTIMER_TIMER
 
+if PM
```
This could be moved to `default y if PM` on the definition of `SOC_SIWX91X_PM_BACKEND_PMGR` instead of being a defconfig override.

asmellby commented on soc/silabs/silabs_siwx91x/siwg917/Kconfig.defconfig:
```diff
@@ -9,4 +9,12 @@ config NUM_IRQS
 config BUILD_OUTPUT_HEX
 	default y
 
+config PM
```
We shouldn't need to redefine these Kconfig symbols, they are defined in subsys/pm/.

jerome-pouiller commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -16,10 +17,19 @@
+#endif
 
-#define DT_DRV_COMPAT silabs_siwx91x_clock
+#define DT_DRV_COMPAT          silabs_siwx91x_clock
```
You probably wanted to remove with line.

jerome-pouiller commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -16,10 +17,19 @@
 #define LF_FSM_CLOCK_FREQUENCY 32768
+#ifdef CONFIG_PM
+#define M4SS_RAM_SIZE_KB (DT_REG_SIZE(DT_NODELABEL(sram0)) / 1024)
+#endif
```
There is no reason to place #ifdef around this definition.

BTW, does it make sense to introduce this intermediate symbol? A local variable with the same name will do the job. so you won't have to pollute the global namespace.

jerome-pouiller commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -35,8 +45,8 @@ static int siwx91x_clock_on(const struct device *dev, clock_control_subsys_t sys
-		RSI_ULPSS_UlpUartClkConfig(ULPCLK, ENABLE_STATIC_CLK,
-					   false, ULP_UART_ULP_MHZ_RC_CLK, 1);
+		RSI_ULPSS_UlpUartClkConfig(ULPCLK, ENABLE_STATIC_CLK, false,
+					   ULP_UART_ULP_MHZ_RC_CLK, 1);
```
I believe the original code was fine.

If you still believe it need to be changes, place place this change in a separated commit (so the reviewer can focus to the important changes).

jerome-pouiller commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -177,6 +187,14 @@ static enum clock_control_status siwx91x_clock_get_status(const struct device *d
+		.m4ss_ram_size_kb = M4SS_RAM_SIZE_KB,
+		.ulpss_ram_size_kb = 4,
+	};
+#endif
```
Remove the `#ifdef` around these declaration (add attribute `__maybe_unused` if the CI complains). So, you will be able to replace `#ifdef CONFIG_PM` by `if (IS_ENABLED(CONFIG_PM))` below.

jerome-pouiller commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -202,6 +220,24 @@ static int siwx91x_clock_init(const struct device *dev)
+	sl_si91x_power_manager_add_ps_requirement(SL_SI91X_POWER_MANAGER_PS4);
+
+	/* Set the clock scaling to performance mode */
+	sl_si91x_power_manager_set_clock_scaling(SL_SI91X_POWER_MANAGER_PERFORMANCE);
```
The comments above each lines just reword the C code. You can probably remove them.

jerome-pouiller commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -212,10 +248,10 @@ static DEVICE_API(clock_control, siwx91x_clock_api) = {
+#define SIWX91X_CLOCK_INIT(p)                                                                      \
+	static struct siwx91x_clock_data siwx91x_clock_data_##p;                                   \
+	DEVICE_DT_INST_DEFINE(p, siwx91x_clock_init, NULL, &siwx91x_clock_data_##p, NULL,          \
+			      PRE_KERNEL_1, CONFIG_CLOCK_CONTROL_INIT_PRIORITY,                    \
```
This is a cosmetics change. Drop or place in a separated commit.

jerome-pouiller commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -24,6 +24,19 @@
 			reg = <0>;
+			cpu-power-states = <&pstate_ps4_standby &pstate_ps4_sleep>;
+		};
+		power-states {
```
Add a blank line above this one.

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 
 set(SISDK_DIR ${ZEPHYR_HAL_SILABS_MODULE_DIR}/simplicity_sdk)
 set(WISECONNECT_DIR ${ZEPHYR_HAL_SILABS_MODULE_DIR}/wiseconnect)
+set(COMMON_DIR     ${ZEPHYR_HAL_SILABS_MODULE_DIR}/simplicity_sdk/platform/common)
```
Either you align all values, either you remove all the repeated spaces.

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -17,6 +18,7 @@ zephyr_compile_definitions(
   SLI_SI917
   SLI_SI917B0
   CLOCK_ROMDRIVER_PRESENT
+  __ZEPHYR_OS_WISECONNECT
```
What don't you use `__ZEPHYR__` (note `__ZEPHYR__` is already used in Wiseconnect)?

fimohame replied:
https://stash.silabs.com/projects/REDPINE/repos/sl_net/pull-requests/6351/overview

The team is suggesting to use the macro as on the PR

Martinhoff-maker commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -160,6 +177,11 @@ if(CONFIG_WISECONNECT_NETWORK_STACK)
     ${WISECONNECT_DIR}/components/sli_si91x_wifi_event_handler/src/sli_si91x_wifi_event_handler.c
     ${WISECONNECT_DIR}/components/sli_wifi_command_engine/src/sli_wifi_command_engine.c
   )
+  zephyr_code_relocate(FILES
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/wireless/ahb_interface/src/rsi_hal_mcu_m4_ram.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/wireless/ahb_interface/src/rsi_hal_mcu_m4_rom.c
+  LOCATION RAM
+  )
```
As I understand, we are now hardlinked with code relocation even if we are not using PM ? If it is only needed in case of PM, It should be guarded by CONFIG_PM IMO.

fimohame replied:
It's used now under PM, but for performance oriented in mere months we might expect those to be moved to RAM on non pm scenarios, for now basis the suggestions I have moved under PM

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -123,7 +140,7 @@ endif() # CONFIG_BT_SILABS_SIWX91X
   zephyr_compile_definitions(
     SLI_SI91X_ENABLE_OS
-    SL_SI91X_SI917_RAM_MEM_CONFIG=1
+    SL_SI91X_SI917_RAM_MEM_CONFIG=2
```
As discussed IRL, we have to rely on DT for this.

BTW, this  fix is not related to PM. I suggest to keep it separated.

Also note `sl_platform_wireless.c` is inconsistent, it should use `.ext_custom_feature_bit_map` rather than `SL_SI91X_SI917_RAM_MEM_CONFIG`. And the values defined in `sl_wifi_device.h` are not used in Zephyr. So, at the end we should be able to drop this definition.

fimohame replied:
SL_SI91X_SI917_RAM_MEM_CONFIG is used across mutliple files in sdk.

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -160,6 +177,11 @@ if(CONFIG_WISECONNECT_NETWORK_STACK)
+  zephyr_code_relocate(FILES
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/wireless/ahb_interface/src/rsi_hal_mcu_m4_ram.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/wireless/ahb_interface/src/rsi_hal_mcu_m4_rom.c
+  LOCATION RAM
```
Indentation (2 spaces).

BTW, do you need to relocate the files or only a few functions. In this case, please filter the relevant symbols. In addition of saving memory, it will be easier to understand why we need that.

Finally, do you need to do that unconditionally or is it only required with `CONFIG_PM`/`CONFIG_SOC_SIWX91X_PM_BACKEND_PMGR`?

fimohame replied:
It's required only during PM, basis your suggestions I have moved as part of PM

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -178,7 +200,44 @@ if(CONFIG_SOC_SILABS_SLEEPTIMER)
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/service/sleeptimer/src/sl_sleeptimer_hal_si91x_sysrtc.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/peripheral_drivers/src/rsi_sysrtc.c
+  ${SISDK_DIR}/platform/service/sleeptimer/src/sl_sleeptimer.c
+  LOCATION RAM
```
Same comments.

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -178,7 +200,44 @@ if(CONFIG_SOC_SILABS_SLEEPTIMER)
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/service/power_manager/src/sl_si91x_power_manager.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/service/power_manager/src/sli_si91x_power_manager.c
+  ${ZEPHYR_BASE}/drivers/gpio/*.c
+  LOCATION RAM
```
Same comments.

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -178,7 +200,44 @@ if(CONFIG_SOC_SILABS_SLEEPTIMER)
+  ${ZEPHYR_BASE}/drivers/gpio/*.c
+  LOCATION RAM
+  )
+endif() #CONFIG_SOC_SIWX91X_PM_BACKEND_PMGR
```
Space after `#`

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -19,6 +19,9 @@
 #endif
+#ifdef CONFIG_PM
+#include "sl_si91x_power_manager.h"
+#endif
```
Remove `#ifdef` (note that there is `#ifdef` around `rsi_ble_common_config.h` because it will break the compilation otherwise)

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -57,7 +61,7 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
 		boot_config->ext_custom_feature_bit_map |= SL_SI91X_EXT_FEAT_352K_M4SS_320K;
 	} else {
-		 k_panic();
+		k_panic();
```
Fine, but keep cosmetics change in a separated commit.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -196,7 +200,10 @@ static int siwg917_nwp_init(void)
+#ifdef CONFIG_PM
+	sl_wifi_performance_profile_t performance_profile = {.profile =
+								     DEEP_SLEEP_WITH_RAM_RETENTION};
+#endif
```
The prefered style is:
```
	sl_wifi_performance_profile_t performance_profile = {
		.profile = DEEP_SLEEP_WITH_RAM_RETENTION
	};
```

Then, you can change `#ifdef CONFIG_PM` in `if (IS_ENABLED(CONFIG_PM))`.

@ArunmaniAlagarsamy2710, watch this change. It will conflict with the Power Save PR.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -223,9 +239,9 @@ Z_ISR_DECLARE(74, 0, IRQ074_Handler, 0);
 		.flags = ISR_FLAG_DIRECT,
 		.func = &siwg917_nwp_stack[sizeof(siwg917_nwp_stack) - 1],
-	};
+};
```
cosmetics (BTW, I believe the original code is fine).

jerome-pouiller commented on soc/silabs/silabs_siwx91x/Kconfig:
```diff
@@ -19,6 +20,12 @@ config SOC_SILABS_SLEEPTIMER
+	bool
+	select SOC_SILABS_SLEEPTIMER
+	help
+	  Implement PM using sl_power_manager service from Gecko SDK
```
Does it make sense to introduce a new symbol? Maybe `CONFIG_PM` is sufficient?

fimohame replied:
zephyr_sources_ifdef(CONFIG_SOC_GECKO_PM_BACKEND_EMU    soc_power.c)
zephyr_sources_ifdef(CONFIG_SOC_GECKO_PM_BACKEND_PMGR   soc_power_pmgr.c)
zephyr_sources_ifdef(CONFIG_SOC_SIWX91X_PM_BACKEND_PMGR soc_siwx91x_power_pmgr.c) 

basis the ref of earlier implementation, I have made similar changes

jerome-pouiller commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -56,8 +58,12 @@ struct gpio_siwx91x_port_data {
 	/* port ISR callback routine address */
 	sys_slist_t callbacks;
+	/* stores the direction of each pin */
+	uint8_t pin_direction[MAX_PIN_COUNT];
```
I believe we need to store boolean. Maybe we can save a few bytes if we replace `uint8_t[]` by `uint16_t` or `uint32_t` (which is the type of `port_pin_mask`).

[EDIT]
An alternative would be to store all the `gpio_flags_t` in this field.

fimohame replied:
Upon this change we might have to bring in pm action changes for uart driver as well.
with the earlier implementation with just the pin direction, we were able to see prints.

as discussed going with the existing code, will take it up when adding pm callbacks for all drivers

jerome-pouiller commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -56,8 +58,12 @@ struct gpio_siwx91x_port_data {
+	uint8_t pin_direction[MAX_PIN_COUNT];
 };
 
+static int gpio_siwx91x_port_pm_action(const struct device *port, enum pm_device_action action);
```
Is this declaration useful?

jerome-pouiller commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -295,6 +301,34 @@ static inline int gpio_siwx91x_init_port(const struct device *port)
+		for (int pin = 0; pin < MAX_PIN_COUNT; ++pin) {
+			if (config->common.port_pin_mask & BIT(pin)) {
+				data->pin_direction[pin] =
+					sl_si91x_gpio_get_pin_direction(config->hal_port, pin);
```
Maybe it would make sense to relocate than in `gpio_siwx91x_pin_configure()`. (So, you could store the flags rather than only the direction).

fimohame replied:
Replied in the previous comment.

Martinhoff-maker commented on drivers/clock_control/clock_control_silabs_siwx91x.c:
```diff
@@ -16,10 +17,19 @@
 #define DT_DRV_COMPAT          silabs_siwx91x_clock
 #define LF_FSM_CLOCK_FREQUENCY 32768
+#ifdef CONFIG_PM
+#define M4SS_RAM_SIZE_KB (DT_REG_SIZE(DT_NODELABEL(sram0)) / 1024)
```
I was thinking that we might have a problem because of the 1 KB that we lost for the NWP at the start of the RAM and the fact that we declare only 255 KB of RAM size for sl_si91x_power_manager(). After digging in configure_ram_memory() function in the sli_si91x_power_manager(), I've found that, in fact, it's having 256 KB of RAM size that might create a problem no ? Isn't it a bug in wiseconnect ?
```
if (config->m4ss_ram_size_kb < RAM_256_KB) {
    // If M4SS RAM Size is less than 256 KB, bit 8 is set.
    *m4ss_ram |= RAM_BANK_8;
  }
```
if we have m4ss_ram_size_kb = 256 KB, we didn't set *m4ss_ram at all ? then ram retention is null I think ?

Martinhoff-maker replied:
It's apparently done on purpose, sorry for this one  :)

jerome-pouiller commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -384,7 +419,8 @@ static DEVICE_API(gpio, gpio_siwx91x_common_api) = { };
 		status = sl_si91x_gpio_driver_enable_clock(                                        \
-			COND_CODE_1(DT_INST_PROP(idx, silabs_ulp), (ULPCLK_GPIO), (M4CLK_GPIO)));  \
+			COND_CODE_1(DT_INST_PROP(idx, silabs_ulp), (ULPCLK_GPIO),                  \
+			(M4CLK_GPIO)));     \
```
cosmetics.

jerome-pouiller commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -361,16 +395,17 @@ static DEVICE_API(gpio, gpio_siwx91x_api) = {
 	irq_enable(DT_IRQ_BY_IDX(node_id, idx, irq));
 
-static DEVICE_API(gpio, gpio_siwx91x_common_api) = { };
+static DEVICE_API(gpio, gpio_siwx91x_common_api) = {};
```
cosmetics (in addition, the original code is correct)

jerome-pouiller commented on west.yml:
```diff
@@ -230,7 +230,7 @@ manifest:
         - hal
     - name: hal_silabs
-      revision: 389726f350880238b9a1034f575ffd46c4309827
+      revision: pull/103/head
```
Ideally, this commit should got first and also contains the required changes in CMakeLists.txt to not break the existing compilation.

jerome-pouiller commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -24,6 +24,19 @@
+				power-state-name = "suspend-to-idle";
+				min-residency-us = <100000>;
+				exit-latency-us = <3000>;
+			};
```
Can you also add the `pm` tag to `boards/silabs/radio_boards/siwx917_rb4338a/siwx917_rb4338a.yaml` and `boards/silabs/radio_boards/siwx917_rb4342a/siwx917_rb4342a.yaml`. So, the required tests will be included in the CI.

fimohame replied:
until pm callbacks are implemented for all the drivers we wouldn't be registering it in the cpu state.
 will keep this in mind.

Martinhoff-maker commented:
How do you test the feature ? because i've got some problems when trying to compile the samples or the test related to PM. I think it comes from the link with Wifi ?

jerome-pouiller commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+			if (sli_si91x_config_clocks_to_mhz_rc() != 0) {
+				LOG_ERR("Failed to configure clocks for sleep mode");
+			} else {
+#if SL_WIFI_COMPONENT_INCLUDED
```
You probably want to rely on a Zephyr symbol (`CONFIG_WISECONNECT_NETWORK_STACK`?).

jerome-pouiller commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+
+LOG_MODULE_REGISTER(siwx91x_pm);
+
+extern uint32_t frontend_switch_control;
```
I suggest to add a getter (in wiseconnect) around this variable rather than exposing it globally.

fimohame replied:
It's part of wifi init structure, and team has requested no patch for it.

jerome-pouiller commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+	default:
+		/* Unsupported power state */
+		LOG_ERR("Unsupported power state %d", state);
+		irq_unlock(0);
```
Why this line is required?

jerome-pouiller commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+	/* Set PRIMASK */
+	__disable_irq();
+	/* Set BASEPRI to 0. */
+	irq_unlock(0);
```
I believe you need to restore the original IRQ mask before to the leave the function:
```
    key = irq_lock();
    __disable_irq();
    __set_BASEPRI(0);
    __ISB();
    ...
    __enable_irq();
    irq_unlock(key);
```

fimohame replied:
I'm seeing functionality issues with the above suggested ways, might have to debug on it.

with the existing code it works as expected.

asmellby replied:
`irq_lock()`/`irq_unlock()` are implemented using BASEPRI on Cortex-M. There is a call to `irq_lock()` in idle.c prior to calling `pm_system_suspend()`, causing BASEPRI to be set. The problem with BASEPRI is that it blocks interrupts from waking you from WFI (sleep). Therefore, it is needed to set PRIMASK using `__disable_irq()`, and then release BASEPRI by forcing it to 0 before going to sleep. Upon waking up, PRIMASK is released to cause ISRs to run.

See the `cortex_m/cpu_idle.c` implementation of `arch_cpu_idle()` for an equivalent implementation. There is also a comment in `idle.c`:

> Note weird API: k_cpu_idle() is called with local CPU interrupts masked, and returns with them unmasked.  It does not take a spinlock or other higher level construct.

So the lack of remembering the BASEPRI level to re-apply after wakeup is also expected.

jerome-pouiller commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+		sli_si91x_m4_ta_wakeup_configurations();
+	} else {
+		/* Device is not ready to sleep; perform necessary actions if required. */
+	}
```
You can probably save some level of indentation by testing exceptions first:
```
    if (!sl_si91x_power_manager_is_ok_to_sleep()) {
        goto out;
    }
    ...
        if (sli_si91x_config_clocks_to_mhz_rc()) {
            LOG_ERR("Failed to configure clocks for sleep mode");
            goto out;
        } 
    ...

out:
    __enable_irq();
```

fimohame replied:
addressed

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -40,20 +46,31 @@ zephyr_include_directories(
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/service/clock_manager/src/sli_si91x_clock_manager.c
   ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_calendar.c
   ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_driver_gpio.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_i2s.c
```
More I think to this code and more I think we going in the wrong direction. This is definitely  not a good idea to blindly import the code from WiseConnect. We have to identify what code is required for what driver and compile it only if required.

We already made some mistakes with `rsi_wwdt.c`, `rsi_pwm.c`, etc... We will fix them later.

fimohame replied:
I have addressed it now..

fimohame commented on soc/silabs/common/soc_siwx91x_power_pmgr.c:
```diff
@@ -0,0 +1,93 @@
+
+extern uint32_t frontend_switch_control;
+
+void sli_si91x_m4_ta_wakeup_configurations(void);
```
I have added a patch for it.

jerome-pouiller commented on drivers/gpio/gpio_silabs_siwx91x.c:
```diff
@@ -56,9 +58,39 @@ struct gpio_siwx91x_port_data {
 	/* port ISR callback routine address */
 	sys_slist_t callbacks;
+	/* stores the direction of each pin */
+	uint16_t pin_direction[MAX_PIN_COUNT];
```
Maybe we want to save ~100bytes if `CONFIG_PM` is not set?

fimohame replied:
On that note the following would be the changes
#if defined(CONFIG_PM)	
	/* stores the direction of each pin */
	uint16_t pin_direction[MAX_PIN_COUNT];
#endif

#if defined(CONFIG_PM)
static int gpio_siwx91x_port_pm_action(const struct device *port, enum pm_device_action action)
{}
#endif

static inline int gpio_siwx91x_init_port(const struct device *port){

#if defined(CONFIG_PM)
	return pm_device_driver_init(port, gpio_siwx91x_port_pm_action);
#else
    return 0;
#endif
}

If these changes are okay then will take it up

fimohame commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -233,6 +234,8 @@ static int siwg917_nwp_init(void)
 	sl_wifi_device_configuration_t network_config;
 	sl_status_t status;
+	__maybe_unused sl_wifi_performance_profile_t performance_profile = {
+		.profile = DEEP_SLEEP_WITH_RAM_RETENTION};
```
moving the closing }; to the next line throws a compliance error,could be coding standard requires the initialization to be written on a single line.

fimohame commented on soc/silabs/silabs_siwx91x/siwg917/soc.c:
```diff
@@ -10,10 +10,28 @@
 #include "em_device.h"
+#ifdef CONFIG_WISECONNECT_NETWORK_STACK
+#include "sli_siwx917_soc.h"
+#endif
```
In the case of Config PM not enabled, it throws an build failure.

fimohame commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -180,5 +196,41 @@ if(CONFIG_SOC_SILABS_SLEEPTIMER)
+    DEBUG_ENABLE
+    DEBUG_UART
+    ENABLE_DEBUG_MODULE
+    SLI_WIRELESS_COMPONENT_PRESENT
```
The macros are from sdk, I suppose we shouldn't be providing it from Kconfig. correct me if I'm wrong

fimohame replied:
DEBUG_ENABLE DEBUG_UART macro's are required as upon wakeup they are reintiliazed in the powermanager sdk file

fimohame committed:
soc: silabs: siwx91x: Add configurable power profile support via DeviceTree

This commit adds support for configuring the power/performance
profile for the siwx91x device using a generic
'power-profile' property in DeviceTree.
The property is available for NWP nodes,
allowing flexible selection of power management
profiles per application or board via overlay.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -65,6 +78,16 @@
+		 *   4: DEEP_SLEEP_WITH_RAM_RETENTION
+		 * Default: 0 (HIGH_PERFORMANCE)
+		 */
+		power-profile = <0>;
```
The need for a comment to explain the power profiles suggests that helper macros should be made available. See https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dt-bindings/regulator/silabs_dcdc.h for an example for DCDC. The NWP power profile should likely also be associated with a generic NWP node, rather than the wifi node. This also means that a `silabs,siwx91x-nwp` binding needs to be created in net/wireless (similar to the `silabs,series2-radio` binding).

I.e. it should IMO be something like this:

```
#include <dt-bindings/net/wireless/silabs_siwx91x_nwp.h>

nwp: nwp {
    compatible = "silabs,siwx91x-nwp";
    power-profile = <SILABS_SIWX91X_NWP_HIGH_PERFORMANCE>;

    bt_hci0: bt_hci {
        compatible = "silabs,siwx91x-bt-hci";
        /* ... */
    };

    wifi0: wifi {
        compatible = "silabs,siwx91x-wifi";
        /* ... */
    };
};
```

asmellby commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -25,6 +25,19 @@
 			reg = <0>;
 		};
+
+		power-states {
```
The CPU should make use of compatible power states, at a minimum the `ps4_standby` state, but also the `ps4_sleep` state if it works out of the box.

fimohame commented on dts/bindings/net/wireless/silabs,siwx91x-nwp.yaml:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2025 Silicon Laboratories Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
```
addressed

jerome-pouiller commented on modules/hal_silabs/wiseconnect/CMakeLists.txt:
```diff
@@ -47,15 +57,26 @@ zephyr_library_sources(
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/core/chip/src/rsi_ps_ram_func.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/cmsis_driver/USART.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/cmsis_driver/UDMA.c
+  ${WISECONNECT_DIR}/components/device/silabs/si91x/mcu/drivers/cmsis_driver/SAI.c
```
Until now we have have included most of the wiseconnect sources files without checking if they were really used by the configuration or not.

This practice does not allow to properly understand (or audit) the architecture of the code and raise concern for the maintenance of the code.

Hence, could you only include in the build the files required by the configuration selected by the user? (For the existing code, I will take of it later in the semester).
